arangod/IResearch/IResearchLink.cpp

```diff
@@ -918,30 +1092,88 @@ Result IResearchLink::commitUnsafeImpl(bool wait, CommitResult* code) {
 
       commitLock.lock();
     }
-
-    auto const lastCommittedTick = _lastCommittedTick;
-
+    _commitStageOne = true;
+    auto const lastCommittedTick = _lastCommittedTickStageOne;
+
+#if ARANGODB_ENABLE_MAINTAINER_MODE && ARANGODB_ENABLE_FAILURE_TESTS
+    TRI_IF_FAILURE("ArangoSearch::ThreeTransactionsMisorder") {
+      std::unique_lock<std::mutex> sync(_t3Failureync);
+      while (_t3NumFlushRegistered < 2) {
+        sync.unlock();
+        std::this_thread::sleep_for(100ms);
+        sync.lock();
+      }
+      _t3CommitSignal = true;
+      LOG_TOPIC("4cb66", DEBUG, TOPIC) << "Commit started";
+    }
+#endif
     try {
       // _lastCommittedTick is being updated in '_before_commit'
       *code = _dataStore._writer->commit() ? CommitResult::DONE
                                            : CommitResult::NO_CHANGES;
     } catch (...) {
       // restore last committed tick in case of any error
-      _lastCommittedTick = lastCommittedTick;
+      _lastCommittedTickStageOne = lastCommittedTick;
       throw;
     }
-
     if (CommitResult::NO_CHANGES == *code) {
       LOG_TOPIC("7e319", TRACE, TOPIC)
           << "no changes registered for arangosearch link '" << id()
-          << "' got last operation tick '" << _lastCommittedTick << "'";
+          << "' got last operation tick '" << _lastCommittedTickStageOne << "'";
 
       // no changes, can release the latest tick before commit
       impl.tick(lastTickBeforeCommit);
-      _lastCommittedTick = lastTickBeforeCommit;
+      _lastCommittedTickStageOne = lastTickBeforeCommit;
+      _lastCommittedTickStageTwo = lastTickBeforeCommit;
       return {};
     }
 
+#if ARANGODB_ENABLE_MAINTAINER_MODE && ARANGODB_ENABLE_FAILURE_TESTS
+    TRI_IF_FAILURE("ArangoSearch::ThreeTransactionsMisorder::StageOneKill") {
+      std::unique_lock<std::mutex> sync(_t3Failureync);
+      // if all candidates gathered and max tick is committed - it is time to
+      // crash
+      if (_t3Candidates.size() >= 3 &&
+          std::find_if(_t3Candidates.begin(), _t3Candidates.end(),
+                       [this](uint64_t t) {
+                         return t > _lastCommittedTickStageOne;
+                       }) == _t3Candidates.end()) {
+        TRI_TerminateDebugging("Killed on commit stage one");
+      }
+    }
+#endif
+    _commitStageOne = false;
+    auto lastCommittedTickStageTwo = _lastCommittedTickStageTwo;
+    // now commit what has possibly accumulated in the second flush context
+    // but will not move the tick as it possibly contains "3rd" generation
+    // changes
+    try {
+      auto secondCommit = _dataStore._writer->commit();
+      LOG_TOPIC("21bda", TRACE, TOPIC)
+          << "Second commit for arangosearch link '" << id() << "' result is "
+          << secondCommit << " tick " << _lastCommittedTickStageTwo;
+      if (!secondCommit) {
+        _lastCommittedTickStageTwo = _lastCommittedTickStageOne;
```

Comment:
```suggestion
        _lastCommittedTickStageOne = lastTickBeforeCommitTwo;
        _lastCommittedTickStageTwo = lastTickBeforeCommitTwo;
```

Reply:
Looks like we could do even better   - we could re-request last tick from engine